File: lib/rin/graphics/DeviceImage.cpp
#include "rin/graphics/DeviceImage.h"

namespace rin::graphics
{
    void DeviceImage::OnDispose()
    {
        
    }

    DeviceImage::DeviceImage(const VmaAllocator& allocator,const VmaAllocation& allocation, const vk::Image& image, const vk::ImageView& view,
        const vk::Extent3D& extent, const ImageFormat& format)
    {
        _allocator = allocator;
        _allocation = allocation;
        _image = image;
        _view = view;
        _extent = extent;
        _format = format;
    }

    ImageFormat DeviceImage::GetFormat()
    {
        return _format;
    }

    vk::Extent3D DeviceImage::GetExtent()
    {
        return _extent;
    }
    vk::Image DeviceImage::GetImage()
    {
        return _image;
    }

    vk::ImageView DeviceImage::GetImageView()
    {
        return _view;
    }
}

File: Rin.Engine/Graphics/DeviceImage.cs
using TerraFX.Interop.Vulkan;

namespace Rin.Engine.Graphics;

/// <summary>
///     GPU Image
/// </summary>
public class DeviceImage : DeviceMemory, IDeviceImage
{
    /// <summary>
    ///     GPU Image
    /// </summary>
    public DeviceImage(VkImage inImage,
        VkImageView inView,
        Extent3D inExtent,
        ImageFormat inFormat,
        Allocator allocator,
        IntPtr allocation, string name) : base(allocator, allocation)
    {
        NativeImage = inImage;
        NativeView = inView;
        Extent = inExtent;
        Format = inFormat;
        Name = name;
    }

    public string Name { get; private set; }

    public ImageFormat Format { get; }

    public Extent3D Extent { get; }

    public VkImage NativeImage { get; }

    public VkImageView NativeView { get; set; }
    
    public override void Dispose()
    {
        Allocator.FreeImage(this);
    }
}